lib/data/wrapper.js

```diff
@@ -309,23 +363,35 @@ const data = {
      * @param {BackendInfo} destBackendInfo - Instance of BackendInfo:
      * Represents the info necessary to evaluate which data backend to use
      * on a data put call.
-     * @param {object} serverSideEncryption - serverSideEncryption
+     * @param {object} sourceBucketMD - bucket metadata of the source
+     * @param {object} destBucketMD - bucket metadata of the destination
      * @param {object} log - Werelogs request logger
      * @param {function} cb - callback
      * @returns {function} cb - callback
      */
     copyObject: (request, sourceLocationConstraintType,
       sourceLocationConstraintName, storeMetadataParams, dataLocator,
-      dataStoreContext, destBackendInfo, serverSideEncryption, log, cb) => {
-        // NOTE: using copyObject only if copying object from one external
-        // backend to the same external backend
-        // and for Azure if it is the same account since Azure copy outside
-        // of an account is async
+      dataStoreContext, destBackendInfo, sourceBucketMD, destBucketMD, log,
+      cb) => {
+        // NOTE: Server side copy should only be allowed:
+        // 1) if source object and destination object are both on aws or both
+        // on azure.
+        // 2) if azure to azure, must be the same storage account since Azure
+        // copy outside of an account is async
+        // 3) if the source bucket is not an encrypted bucket and the
+        // destination bucket is not an encrypted bucket (unless the copy
+        // is all within the same bucket).
         const locationTypeMatch = request.headers[locationHeader] ?
         config.getLocationConstraintType(sourceLocationConstraintName) ===
         config.getLocationConstraintType(request.headers[locationHeader])
         : true;
-        if (locationTypeMatch &&
+        const sourceBucketName = sourceBucketMD.getName();
+        const destBucketName = request.bucketName;
+        const isSameBucket = sourceBucketName === destBucketName;
+        const serverSideEncryption = destBucketMD.getServerSideEncryption();
+        const bucketNotEncrypted = serverSideEncryption
+        === sourceBucketMD.getServerSideEncryption() === null;
+        if (locationTypeMatch && (isSameBucket || bucketNotEncrypted) &&
```

Comment:
I would wrap all of this logic determining whether we can do server side copy in a utility function. That way it will be clearer and can be used for copy part as well. Plus, you can unit test it.